Collision various fixes

TheTermos opened the pull request:
The main goal of this PR is to fix collision detection issues related to floating point precision (e.g. #6886, #9143, #9305, #5966)

Fixes so far:
- fixed false positive collisions altogether.
- fixed boxes smaller than 0.27 clipping into nodes.
- fixed inconsistent object-object collisions
- fixed client ignoring object-localplayer collisions
- fixed step-up glitches.
- significantly reduced average amount of calculations per box in range
- ~~fixed sneak jump causing player clip into nodes~~ another PR
- ~~fixed not being able to sneak down the stairs~~ another PR

**[200206 solution 2]** rewritten AxisAllignedCollision() from scratch.
Now it works on a slightly different principle, instead of a face colliding with anbother one, it's a face colliding with a whole volume of a box.
- so far it seems to work flawlessly without the need for any uncertainty radius for object-node and object-object collisions.
- performance seems improved.
- it wouldn't need the concept of negative time collisions if it was a purely local scenario, however in MT it is commonplace that due to cient-server disagreement about objects' whereabouts their position is set so they intersect, so I implemented negative time collisions, but it's not based on some arbitrary error margin, it occurs only when boxes overlap and exactly as much as needed for them to separate.

Expectedly, there are slight issues with existing old code that sets objects position so they overlap.
- ~~/spawnentity works differently because an entity is placed at player's position.~~ no longer apply
- ~~infrequent glitch with sneak jump - under certain conditions sneak code sets player position so the box is clipping with a node~~ **fixed**
- ~~an old bug in step-up feature that sometimes sets objects position so they clip with nodes, but this PR actually seems to make things better in this case.~~ **fixed**

## To do

This PR is Ready for Review 

## How to test
Go crazy with object-object and object-player collisons.

**Note:** Due to recent changes in localplayer, this PR won't work with 5.1

![collision](https://user-images.githubusercontent.com/55103816/73142682-fc90b900-4091-11ea-944e-13bfd6bb35fc.gif)

paramat commented:
> fix collision detection issues related to floating point precision.

Please could you briefly explain what these issues are?

TheTermos commented:
> Please could you briefly explain what these issues are?

Updated the description with issue ids. 
As to client side issues described above, I don't know if there are any issues posted on github, but these problems have been at least since 0.4.17, I made a forum post some time ago https://forum.minetest.net/viewtopic.php?f=6&t=22255

This PR addresses the same issues as #9327 and more.
The advantage of #9327 is that it's simple, ready to go and can be easily backported, but it's not a complete solution, this one is going to be.

Could you please set the Testing Needed label?

nerzhul commented on src/client/localplayer.h:
```diff
@@ -141,6 +141,8 @@ class LocalPlayer : public Player
 
 	void setCollisionbox(const aabb3f &box) { m_collisionbox = box; }
 
+	aabb3f getCollisionbox() const { return m_collisionbox; } //TheTermos
```
const aabb3f &

TheTermos replied:
Received convention ;) the other one as well, it was already like that elsewhere in the code, but I get the point.

I'm going to modify these, thanks, but this code will keep changing anyway, so I propose let's leave such touches for the review phase, I'll try to get it tidy by then.

TheTermos replied:
changed as requested

nerzhul commented on src/collision.cpp:
```diff
@@ -405,6 +406,18 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 			}
 		}
+#ifndef SERVER
+		if (c_env != 0) {
```
if (c_env) {

TheTermos replied:
changed as requested

paramat commented:
Because 5.2.0 will be released soon i really think this should be merged after release, as fundamental changes to collision is likely to cause unforseen bugs to pop up much later.
Then we will have a whole release cycle for testing and catching any bugs.

TheTermos commented:
> Because 5.2.0 will be released soon i really think this should be merged after release, as fundamental changes to collision is likely to cause unforseen bugs to pop up much later.
> Then we will have a whole release cycle for testing and catching any bugs.

Yes, I intended it for later. 
I also have another quite big modification ready that increases position calculation precision to 64 bit (#9346)  which I think belongs before this one and after 5.2

LoneWolfHT commented:
Can't reproduce the Pyramid or Sneak glitch with this

TheTermos commented:
Updated this PR with another solution, this one I think has the potential to become final.
- fixed false positive collisions altogether.
- fixed inconsistent object-object collisions
- fixed client ignoring object-localplayer collisions
- significantly reduced average amount of calculations per box in range.

The sneak jump issue mentioned in the OP, I think it should be fixed in localplayer.cpp with another PR.

TheTermos commented:
Changed my mind, rewritten LocalPlayer::updateSneakNode() within this PR

I wouldn't normally touch the sneak code, but the old one was entirely incompatible with working collision detection.
I'd be grateful for any opinions on the new sneaking code.

SmallJoker commented:
> but the old one was entirely incompatible with working collision detection.

I'd like to inform you that touching the sneak code is a dangerous thing to do. Last time it caused a holy war or player opinions and the introduction of `LocalPlayer::old_move()` for those who did not like it.

LoneWolfHT commented:
> > but the old one was entirely incompatible with working collision detection.
> 
> I'd like to inform you that touching the sneak code is a dangerous thing to do. Last time it caused a holy war or player opinions and the introduction of `LocalPlayer::old_move()` for those who did not like it.

The current sneak is lifeless and without character, I think it should be pretty easy to change it without too many complaints

TheTermos commented:
> I'd like to inform you that touching the sneak code is a dangerous thing to do. Last time it caused a holy war or player opinions and the introduction of `LocalPlayer::old_move()` for those who did not like it.

@SmallJoker Thanks, I figured that out.
I really wouldn't touch the sneak code if it wasn't necessary. Is there any record of that holy war conversation so I can get an idea what to expect?

I guess we're going to see if this is about making things better or preserving old bugs.

@paramat has put forward a reasonable solution to keep the sneak glitch when old move option is in effect, there are no reasons for me to touch LocalPlayer::old_move(), so that can be done.

paramat commented:
See past issues/PRs related to sneak jump.

My suggestion about altering sneak jump/glitch in new movement code is here https://github.com/minetest/minetest/pull/9365#issuecomment-583770958

There is not much harm in altering sneak jump/glitch in new movement code, as we have the option of old movement code for those who like the old behaviours, inherently they will have to accept the bugs that go with that.

I believe we should remove the sneak jump simulations from new movement code because these are just creating new complexities that have to be maintained into the future, with much resulting headache and restriction of development.
They were added before the option of old movement code was added, so there is less need for them now. Also, those who like sneak jump usually use the old movement code and dislike the new simulations, which are significanly different.
It seems best they are removed before players get used to them and we end up just repeating what happened before because we have to fix something but fixing is incompatible with the highly sensitive sneak jump simulations. In fact this is what is happening now.

I made an issue about this a while ago #7107

paramat commented:
Actually, it may be best you do not try to understand the sneak jump controversy by reading old issues and PRs, as the discussions and arguments are very long, unpleasant, and will not make it clear.
Instead i will write a summary of what you need to know in this thread soon.

At the least, i think that we should consider the sneak jump/ladder/glitch simulations in new movement code to be lower priority than improvements and bugfixes. I think this is obvious because otherwise we will then be trapped in the same way the old movement code trapped us.

I think we should make it clear that the only sneak behaviour officially supported in new movement code is sneaking. Any other unintended sneak related behaviour, either old or newly arising, is not supported and may change.

TheTermos commented:
@paramat Thank you,

I haven't concerned myself with sneak glitch so far, so even though old_move is unchanged, it probably won't work because my collision detection is too good at unclipping clipping boxes.
But all it takes is to add old collision.* as a separate file and make old_move use that, which should be trivial.

The 'new' sneak code, I tried to reproduce it as closely as possible, except sneaking down the stairs because I couldn't see why anyone would want to be able to sneak up the stairs but not down.
But of course changes are possible, given there's demand.

TheTermos commented:
Got a query about sneaking.
Do you guys know if there are any specific reasons why player can't step down a full node while sneaking?

I can easily make it so he can step down but only as far as he'll be able to sneak back up, that would improve the flow of sneaking without taking away anything important I guess.

paramat commented:
#5438 is a list of current new move sneak-related bugs.
#7725 might be related.
> player can't step down a full node while sneaking?

I do not understand what you mean, surely sneaking is meant to stop a player falling off an edge even if the fall is 1 node?

LoneWolfHT commented:
Make it a clientside option disabled by default if you do it. I can see people liking it and not liking it

TheTermos commented:
> I do not understand what you mean, surely sneaking is meant to stop a player falling off an edge even if the fall is 1 node?

Yes, that's what I meant. One node height difference can be seen as normal MT terrain rather than a ledge, with this PR all it takes to modify that behavior is just one parameter change.
But if that's intentional, this is how the PR works now and no action is required.

> Hmmm, by 'sneak back up' do you mean 'sneak back up without jumping'? I'm up for testing if so, that would be really nice

This PR works like that, but only up to stepheight height difference, a full node has to be jumped back up.
What you describe can be done relatively easily, but requires fixing step-up code, which is currently  the more likely to cause glitches the greater the stepheight.

LoneWolfHT commented:
Maybe limit it to 1 node if the stepheight is > 1
I don't think builders will like falling off 1-node ledges unless stepheight allows them to walk back up without jumping. I guess if this was added as a setting disabled by default we could test

TheTermos commented:
> I don't think builders will like falling off 1-node ledges unless stepheight allows them to walk back up without jumping.

Hmm, maybe.
I think I'll just stick to recreating the 'new' sneak as close as possible unless otherwise requested.

paramat commented:
Yes it is intentional that sneaking stops falling off any edge, whatever the fall distance is.
See this list of current new move sneak-related bugs https://github.com/minetest/minetest/issues/5438#issuecomment-350406456 , sneak allowing stepping-down by a distance of less than 1 node onto a nodebox below is seen as a minor bug.

TheTermos commented:
> See this list of current new move sneak-related bugs [#5438 (comment)](https://github.com/minetest/minetest/issues/5438#issuecomment-350406456) , sneak allowing stepping-down by a distance of less than 1 node onto a nodebox below is seen as a minor bug.

Yes, very interesting.
It's a fascinating process to watch, when bugs go out of control for so long that they take control of the project. Looks like there weren't enough of the right people at the right time on the dev team to put a stop to that.

The criteria of bugginess have become arbitrary and these threads are not informative enough for me to identify all the many little glitches that need reproducing, so I've got a better bad idea. I'm going to revert my changes in localplayer and keep calling old collision while sneaking, regardless old or new. That should hopefully keep everyone happy and let me stay away from the danger zone.

paramat commented:
> The criteria of bugginess have become arbitrary and these threads are not informative enough for me to identify all the many little glitches that need reproducing

Do not worry, it is simple:
* Movement behaviour when 'old movement code option' is selected must remain unchanged, that is the only danger zone.
* Movement behaviour when 'new movement code option' is selected is free to change as long as obvious and officially supported movement behaviour is preserved.
In terms of sneak, 'officially supported movement behaviour' is simply not falling of an edge, nothing else.

So you are free to alter the behaviour of the sneak jump/ladder simulations in new movement code, and the behaviour of not officially supported sneak-related behaviour, because bugfixes and improvements are higher priority than those. It has to be this way otherwise we are letting ourselves be trapped by not officially supported behavour, which makes no sense, we do not do this anywhere else in code.

We should make it clear to players that any not officially supported movement behaviour in the 'new movement code option' is likely to change, to avoid the previous controversy.
For example: New movement code has created some new unforseen sneak-related behaviours, such as new types of sneak-jump, but we will not support and preserve these. Players should be clearly warned to not get used to not officially-supported movement behaviours.

Thinking we have to instantly preserve every not officially supported unusual movement behaviour in the 'new movement code option' would be insanity, a huge headache and will prevent movement code improving.
> keep calling old collision while sneaking, regardless old or new.

So this is unnecessary and the wrong approach. Feel free to improve things as long as sneaking stops falling of an edge.

I will write my summary soon to help you understand.

paramat commented:
Original movement code
----------------------

The original movement code had the 'sneak' feature that prevents falling off an edge when holding the sneak key.
However it was coded in a way that caused unintended behaviours, one of which came to be called the 'sneak glitch':

Imagine a flat wall in front of you, with one node removed from the wall to form a cavity, where the cavity floor is 1 node up from ground level.
By jumping up at the cavity while sneaking, the player is able to sneak on the cavity edge, despite there not being a 2-node high space to stand in.

Additionally, if cavities are arranged beside each other but staggered, this sneak-jump can repeat and results in a rapid vertical climb, where officially the structure should not be climbable, this unintended behaviour came to be called the 'sneak ladder'.

This was discussed in 2013 in https://github.com/minetest/minetest/pull/760
2 core devs considered it unintended and overpowered behaviour. It is interesting to see the controversy occurs here too, many players had got used to these 2 unintended behaviours and wanted them to continue.

Sneak glitch option
-------------------

In that thread celeron55 suggested that this fix be made optional, controlled by a player physics override, this was implemented in https://github.com/minetest/minetest/pull/1035
Unfortunately the default for the 'sneak glitch' option was set to 'true', which probably caused many new players to get used to these unintended behaviours, leading to later problems.

Sneak jump
----------

Another unintended behaviour was possible that came to be called the 'sneak jump':
By sneaking while jumping it was possible to pass over a 2-node high wall, despite player jump ability only being 1 node.
Again many players got used to this, and built maps that relied on the ability.

New movement code
-----------------

As far back as 2012 it was noticed that fall damage could be avoided in some situations if sneak is held, see https://github.com/minetest/minetest/issues/329
In 2017 this was fixed in https://github.com/minetest/minetest/pull/5327
To make that possible, the 'sneak jump' and 'sneak ladder' behaviours were no longer possible. However, 'sneak glitch' was still possible.

Because this was a rewrite of movement code, this change was not made optional.
This caused 'the sneak apocalypse'. 4 years on many more players were now used to the unintended behaviours and had built maps relying on them.
During the following arguments one player started an extremely hostile fork of MT.

Sneak jump and sneak ladder simulations
---------------------------------------

So then much time was spent coding simulations of 'sneak jump' and 'sneak ladder' in the new movement code. The 'sneak glitch' physics override then became the option of using these simulations, but this time the default was set to 'false'.
These simulations were close but unavoidably non-identical, players tend to become used to very specific responses to controls so most found the simulations unusable.

Old movement code option
------------------------

So eventually we re-added the old movement code and made it a option, controlled by the 'new move' physics override, default 'true'.
This was originally intended to be a temporary option while we tried to improve the simulations in the new movement code. However eventually it became clear that many players found the simulations unusable, so we decided to make this option permanent.

paramat commented:
Keep in mind that most of #5438 was written before i re-added old movement code as an option in #5519 . So most of that thread was written at a time where there was intense concern about the simulations of 'sneak jump' and 'sneak ladder' in the new movement code, and intense pressure for those simulations to be as faithful as possible.
That is no longer the case, those who like the 'sneak jump' and 'sneak ladder' of old movement code use the old movement code option.

TheTermos commented:
Thank you,
However it still isn't obvious to me at all why sneak preventing falling off an edge **of any height** is officialy supported behavior.

It feels inconsistent, as you can sneak up, but not down, this is particularly puzzling when sneaking the stairs, that's why initially I considered it a bug, and chances are most people unfamiliar with MT's past see it that way.
My guess is there's some history involved and that behavior feels natural to the oldtimers who may see it as build-sneak, or parkour-sneak, but on sneak-sneak that doesn't make a whole lot of sense to me.

An example, in one of my mods zombies can be snuck upon from behind, because they can only hear players moving at normal speed. But if there's any edge in between, even a fraction of a node high, the sneak has to be released and a short distance walked normally - you make noise, stealth factor is out of the window and you get your brains eaten.

Maybe this is supposed to be something else and the name 'sneak' is misleading, maybe it should be renamed according to what it's supposed to be, but then such a never-down rule might be fine in a game context, but seems awfully specific for an engine.

Or maybe what's really needed is a climb feature?
It could work so it prevents player from falling whenever there's an exposed top nodebox face adjacent to player's colbox.
Notice it would allow for node ladders, 2 node jump and then some, would make perfect sense on the assumption players have hands (holding onto the ledges) and all that without buggy clipping.

rubenwardy commented:
I think it's fine to change behaviour such as sneaking not falling off any edges, but it should not be done in a bug fix PR and should gain community feedback first. I don't have any particular opinion on that behaviour, I just want the bugs to be gone

paramat commented:
> why sneak preventing falling off an edge of any height is officialy supported behavior.

Allowing falling a small height might possibly be acceptable if we discuss and decide on that. However i expect it would be controversial and unpopular. I think any change to this would need very good reasons and i do not see any.

The current 'not falling any height' is just what players are used to.
It makes sense to me, as a player may wish to hang on to an edge above a small fall to help reach/dig/place something, or to help jumping up onto something. The building and parkour uses seem to me to be the priority, sneak is not primarily a stealth thing in MT.
What MT players are used to, and what worlds are built for, is more important than any assumptions or expectations newbies may have.

The 'inconsistency' (i do not see it as one) does not seem a problem to me, sneak stops falling off an edge, so that always happening seems right. It is like a latch, you go up but cannot come down.
There is also the question of how far a fall to allow when a players's step or jump height is altered. Jump height is the result of a complex to calculate combination of the per-player physics overrides 'jump' and 'gravity' so it gets complex. Seems better to just keep it simple and stop falling off every edge. If a game does not allow jumping up 1 node then a player will not want to fall 1 node using sneak.

LoneWolfHT commented:
If it allowed sneaking off a drop <= stepheight I could see it working quite well, probably best in a separate PR like Ruben said though

TheTermos commented:
> The current 'not falling any height' is just what players are used to.

Stopping on every edge is how it works in many games, but most of them have smooth slopes and very little deliberately placed edges. In MT however slopes are made of edges, that makes the feature largely unusable for the intended purpose (as the name 'sneak' suggests, I'm sure some minetesters could have found many exotic uses for that)

If you're worried about complexity, you might compare localplayer.cpp in this PR to master and see for yourself which one is more complex.

> If it allowed sneaking off a drop <= stepheight I could see it working quite well, probably best in a separate PR like Ruben said though

That's how it works now, you can still try it out. By the looks of it I'll have to revert all changes to localplayer soon.

paramat commented:
> In MT however slopes are made of edges, that makes the feature largely unusable for the intended purpose (as the name 'sneak' suggests

No ... =) 'Sneak' is completely usable because it works as MT players expect it to, are used to, and according to its function in MT. In your statement you have redefined the purpose based on a literal interpretation of the word or how you think it should function.
The word 'sneak' was only used because MT was inspired by MC, the function in MT is (or was) not falling off any edge (although in new movement code the player will actually fall down if the node below is a nodebox, see the issue, this is considerd a minor bug).

However ...
> If it allowed sneaking off a drop <= stepheight

This is probably fine, although it would be wise to have some player input on this first.
It is also simple because stepheight will always be less than jump height, so we will not need to calculate jump height.

TheTermos commented:
'redefined based on literal interpretation' is probably an oxymoron.
I find the above reasoning all kinds of wrong assuming MT is supposed to be an engine, but even that is debatable. 

Luckily all that may just have become irrelevant, I think I have found another way to fix collision detection without modifying localplayer, I'll be reverting all changes to it shortly.
If you guys like it broken, who am I to argue ;)

I might make another PR decicated to localplayer move some day, if anyone's interested.

LoneWolfHT commented:
If you're talking about the stepheight method I'm interested

TheTermos commented:
Fixed step-up feature.

Well, that's probably too strong of a word, the algo is actually quite clever, it just has an inherent limitation -  it guarantees glitches whenever stepheight exceeds collisionbox height.

I only made sure minimum of stepheight and colbox height is used.
If ever it is decided it's necessary that step-up works with greater values than colbox height, the algo will have to be rewritten from scratch.

paramat commented:
> assuming MT is supposed to be an engine, but even that is debatable.

=) Yes, MT is a game engine but it is orientated towards 'Minecraft-like' or 'Minetest-Game-like' cubic voxel games and is not extremely flexible (we do not have the core dev time to make it so).
The 'MT is supposed to be an game engine' argument comes up sometimes to try to justify inappropriate features but it is actually a somewhat invalid argument.

paramat commented:
> I find the above reasoning all kinds of wrong assuming MT is supposed to be an engine,

I do not understand, why is it wrong?

If you mean that your response to me stating that sneak should stop falling off every edge is that MT is a game engine and therefore sneak behaviour should be more configurable, then, i have no objection to that, but that is a different discussion and something for the future (as rubenwardy wrote above).
For now, and in this collision bugfix PR, there is only one sneak behaviour (in new movement code), so it should preserve as closely as possible the official behaviour that users are used to and what they and their worlds rely on, this seems obvious.

I might even agree with you that sneak should ideally behave differently, or that it does not make sense in terms of 'edges', or that it does not match the word 'sneak' well. However, all this is irrelevant personal preference, altering basic official sneak behaviour can enrage huge numbers of players, break many worlds and bring about the second apocalypse.

But, i expect that allowing a fall of <= player step_height will not be very controversial.
>  it guarantees glitches whenever stepheight exceeds collisionbox height.

Referring to player collisionbox height i assume?

TheTermos commented:
> If you mean that your response to me stating that sneak should stop falling off every edge is that MT is a game engine and therefore sneak behaviour should be more configurable, then, i have no objection to that, 

Exactly. Engines are meant to be versatile, that is implement as few specific features themselves as possible, in order to enable users to implement as many specific features as they need. If a specific feature can't be avoided, its' purpose at least should be clearly defined, however in this case the best I could get were worries about some subgroup of players having potentially gotten used to it.
That kind of obscure community emerged features got their place in a game as a service model, much less in an engine.

But as i said, chances are I'll be able to avoid modifying localplayer move entirely, so this is best discussed some other time.

> Referring to player collisionbox height i assume?

Referring to any moving box, player or object.

Say we've got a size 0.4 cube, if we wanted it to be able to walk stairs and allowed it stepheight of 0.55, there would be glitches.

This modification is entirely in collision.cpp, nothing player specific.

runsy commented:
Related to this bug:
[https://forum.minetest.net/viewtopic.php?f=9&t=22245&start=700#p367852](https://forum.minetest.net/viewtopic.php?f=9&t=22245&start=700#p367852)

TheTermos commented:
Here's the deal with step-up.
Consider a small box of height < 1.0 and stepheight of 1.6. It will be able to step up a full node or an oversized nodebox, but won't be able to step up a full node if there's a slab on top of it.
This is clearly an improvement over the current step up code, which would glitch into the slab in that scenario, or if the box hit a tree it would glitch into the tree.
This is the best I could do without rewriting step up code from scratch.

**a technical question**
Is there a way to cleanly revert the commit that done changes to localplayer? I'm using the web gui exclusively atm.
Or maybe I should make another PR to keep things clean, with one commit per file with nice comments, and close this one, which has become quite commit heavy?

TheTermos commented:
Now I consider this finished, 
that is I don't intend to make any further modifications unless there are problems reported.
It also should be totally uncontroversial, all changes to localplayer.cpp have been undone.

Would be very helpful if people who know how old move is supposed to work tested it to see if modified collision detection interferes in any way with it (it shouldn't).

SmallJoker commented on src/client/localplayer.h:
```diff
@@ -141,6 +141,8 @@ class LocalPlayer : public Player
 
 	void setCollisionbox(const aabb3f &box) { m_collisionbox = box; }
 
+	const aabb3f& getCollisionbox() const { return m_collisionbox; } //TheTermos
```
`const aabb3f &getCollisionbox`

redundant comment

TheTermos replied:
done

SmallJoker commented on src/collision.cpp:
```diff
@@ -61,118 +63,105 @@ struct NearbyCollisionInfo {
+		distance = relbox.MaxEdge.X - relbox.MinEdge.X;
+		//distance = distance >= 0.03f ? distance - 0.03f : distance;
+
+		*dtime = distance>=0 ? std::abs(distance / speed.X) : -std::abs(distance / speed.X);
```
Please put spaces around binary operators: `distance >= 0`

TheTermos replied:
done

SmallJoker commented on src/collision.cpp:
```diff
@@ -61,118 +63,105 @@ struct NearbyCollisionInfo {
+		if (distance > inner_margin) {
+			if (*dtime <= dtime_max) {
+				if (((speed.X > 0) && (staticbox.MaxEdge.X > movingbox.MaxEdge.X)) ||
+					((speed.X < 0) && (staticbox.MinEdge.X < movingbox.MinEdge.X)))
```
Bracket of the next line should be moved here. Also: Superfluous parentheses.

TheTermos replied:
done

SmallJoker commented on src/collision.cpp:
```diff
@@ -61,118 +63,105 @@ struct NearbyCollisionInfo {
+						(std::max(movingbox.MaxEdge.Z + speed.Z * time, staticbox.MaxEdge.Z)
+							- std::min(movingbox.MinEdge.Z + speed.Z * time, staticbox.MinEdge.Z)
+							- relbox.MinEdge.Z < 0)
+						) return COLLISION_AXIS_X;
```
New line before `return`
EDIT: Same thing below.

TheTermos replied:
done

SmallJoker commented on src/collision.cpp:
```diff
@@ -61,118 +63,105 @@ struct NearbyCollisionInfo {
+		time = std::max(*dtime, 0.0f);
+
+		if (distance > inner_margin) {
+			if (*dtime <= dtime_max) {
```
Conditions should be combined to reduce the amount of indents below.

TheTermos replied:
One of these ifs has an else, so in this case combining would actually make things worse.

SmallJoker commented on src/collision.cpp:
```diff
@@ -405,6 +394,18 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
+				v3f lplayer_pos = lplayer->getPosition();
+				lplayer_collisionbox.MinEdge += lplayer_pos;
+				lplayer_collisionbox.MaxEdge += lplayer_pos;
+				cinfo.emplace_back(false, true, 0, v3s16(), lplayer_collisionbox);
```
So how does the player not collide with their own collision box?

TheTermos replied:
Here: 398 `if (self && c_env)`
Localplayer doesn't set the self so took advantage of that.

SmallJoker commented on src/collision.cpp:
```diff
@@ -61,118 +63,105 @@ struct NearbyCollisionInfo {
+				{
+					if (
+						(std::max(movingbox.MaxEdge.Y + speed.Y * time, staticbox.MaxEdge.Y) 
+							- std::min(movingbox.MinEdge.Y + speed.Y * time, staticbox.MinEdge.Y) 
```
2x trailing whitespace

TheTermos replied:
done

TheTermos commented:
> I'd like to apply this PR onto the current master, but merge "rebase" makes this impossible. Could you please squash & rebase this branch for easier testing?

Sure.
It's just my PC programming experience is so 20th century, I'm new to this git thing, so any advice how to do it easily and preferably using the web interface would speed things up.

SmallJoker commented:
I don't know how, but the today's diff applied just fine. :man_shrugging:
As soon a PR needs a rebase, you'd normally do
```sh
# 1) Save local changes
git stash
# 2) Update from minetest/minetest
git pull upstream/master --rebase
# 3) Solve conflicts (marked with `<<<<<` in files)
# 4) Add the files with solved conflicts and continue
git add solved_conflict_file1.ext solved_conflict_file2.ext ...
git rebase --continue
# 6) When no conflicts are left over: re-apply the local changes
git stash pop
```

So far I don't see any side-effects. Tested cases:

- Normal player movement
- Attached to non-physical entity
- Attached to physical entity, and colliding with another entity

TheTermos commented:
> I don't know how, but the today's diff applied just fine.

Whew!
Maybe it has something to do with the new commit I made today? 
Can we forgo the cosmetics in collision.cpp then, if it works? I'd rather not have to rebase this, never used commandline git before. I tried interactive rebase and force push on a test repo, and it seemed to work, but this is 19 commits, so I'd probably blow the thing up in the process.

> So far I don't see any side-effects. Tested cases:

Could you check if it interferes with old move in any way? I don't know how it is supposed to work in detail, and to be honest I prefer not to, if not necessary.

paramat commented:
I disapprove merging this PR before 5.2.0.

Cahrs commented:
On the topic of sneaking on stairs, as a player who mourned the loss of the original sneak glitch (ever since it became a server setting its become an endangered species due to many new server owners not even knowing of its existence) i have always hated how sneaking along stairs work, and would love to see it fixed. As it is doing things like sneaking to the edge of a slanted roof is a PITA and slowly peeking over the top of one is pretty much down right impossible without sneak-bunnyhopping. The only issue i can see with changing it is parkour maps depending on it, but considering that a lot of parkour maps were also built with the original sneak glitch in mind, i think it would work out fine as long as it stays the same in the legacy sneak code option.

paramat commented:
>  as a player who mourned the loss of the original sneak glitch (ever since it became a server setting its become an endangered species due to many new server owners not even knowing of its existence)

Original sneak-related behaviour is not 'lost', and you can simply contact the server admin to inform them.
You could then try to convince then to enable the 'old movement code', you have a good chance as very many players prefer it.

TheTermos commented:
How did the merge run only after the checks have failed?

sfan5 commented on src/unittest/test_collision.cpp:
```diff
@@ -132,48 +132,48 @@ void TestCollision::testAxisAlignedCollision()
 			v3f v(1./7, 1./7, 1./7);
-			f32 dtime;
-			UASSERT(axisAlignedCollision(s, m, v, 0, &dtime) == 1);
+			f32 dtime = 17.0f;
```
where do these values come from?

TheTermos replied:
dtime parameter has been changed to input/output.
On input it signifies the time left in the step, here it is set as needed for the collision to occur according to speed and distance.

This is a performance optimization, previously calculations were run on all boxes in radius (box extent +1.5n), now boxes are rejected based on distance between faces > dtime*speed

sfan5 commented on src/collision.cpp:
```diff
@@ -574,7 +583,8 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 			}
-			if (std::fabs(cbox.MaxEdge.Y - box.MinEdge.Y) < 0.15f * BS) {
+			//if (std::fabs(cbox.MaxEdge.Y - box.MinEdge.Y) < 0.15f * BS) {
+			if (std::fabs(cbox.MaxEdge.Y - box.MinEdge.Y) < 0.05f) {					//TheTermos make it less stupid after step-up fix
```
why is the `* BS` gone here?

TheTermos replied:
0.15 * BS translates to 15 world centimeters, an extreme value, that's the reason why touching_ground has been useless. 

0.05 should be enough. It relates to calculation precision rather than world dimensions, so it needn't be expressed in BS.

sfan5 commented on src/collision.cpp:
```diff
@@ -419,8 +420,9 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
-	assert(d > pos_max_d);	// invariant
+	f32 d = 0.0f;	// Temporary fix, any nonzero d causes collision glitches, the more the greater it is.
+	// ultimately it has to be determined if any uncertainty is involved, and if it is, eliminated
+	// and d & pos_max_d params removed from function calls.
```
None of this applies anymore does it?

TheTermos replied:
I intended too keep it until the tests gone ok, as a reminder to update the collisionMoveSimple parameter list everywhere it's called, is that ok?

sfan5 replied:
Just make sure to remove it before merge.

TheTermos replied:
removed the comment, the d stays because it's used by wouldCollideWithCeiling() and in touching_ground determination.

sfan5 commented on src/collision.cpp:
```diff
@@ -25,11 +25,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+#include <iomanip>
 
 // float error is 10 - 9.96875 = 0.03125
 //#define COLL_ZERO 0.032 // broken unit tests
```
`COLL_ZERO` should be removed if it is unused now.

TheTermos replied:
Haven't touched any of that, the comment was there before, should I still remove it?

sfan5 replied:
Well you have removed some uses of `COLL_ZERO`, so are the any left?

TheTermos replied:
removed

sfan5 commented on src/collision.cpp:
```diff
@@ -25,11 +25,13 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "serverobject.h"
 #include "util/timetaker.h"
 #include "profiler.h"
+#include <iomanip>
```
what is this needed for?

TheTermos replied:
Probably leftover from my testing, I must have been outputting some text, gonna check and get rid of.

TheTermos commented on src/collision.cpp:
```diff
@@ -61,118 +62,86 @@ struct NearbyCollisionInfo {
-			movingbox.MaxEdge.Z - staticbox.MinEdge.Z
+			movingbox.MaxEdge.X - movingbox.MinEdge.X + staticbox.MaxEdge.X - staticbox.MinEdge.X,						// sum of the widths
+			movingbox.MaxEdge.Y - movingbox.MinEdge.Y + staticbox.MaxEdge.Y - staticbox.MinEdge.Y,
+			movingbox.MaxEdge.Z - movingbox.MinEdge.Z + staticbox.MaxEdge.Z - staticbox.MinEdge.Z,
```
Will keep in mind for the future, but now that there are no typos, let's not make a new one while rearranging.

TheTermos commented:
So, are we ready yet?

@nerzhul requested some changes months ago and seems to have gone afk, hope that's not a problem., the changes are long done and forgotten.

sfan5 commented:
No that's not a problem, this PR now just needs another core developer to (review and) approve it.

TheTermos commented:
Yes, the differences become more apparent when working with entities.

runsy commented:
Finally merged!!! :-D

KaylebJay commented:
Yes! Awesome! :D

TheTermos commented:
Stumbled upon something weird.

Was testing player-object collisions, and player would occasionally end up moved to infinity.
Done some debug, and it turns out localplayer.cpp sometimes passes oddly slow velocity components to collisionMoveSimple()

inserted the following right in the beginning of collisionMoveSimple
```
if (speed_f->X != 0.0 && speed_f->X < 1.0e-11 && speed_f->X > -1.0e-11) 
		warningstream << "input" << std::scientific << speed_f->X << '\n';
```	

example result:
```
2020-04-21 20:17:42: WARNING[Main]: input1.730467e-27
2020-04-21 20:17:43: WARNING[Main]: input-7.042432e-19
2020-04-21 20:17:43: WARNING[Main]: input4.193221e-37
2020-04-21 20:17:50: WARNING[Main]: input-1.928631e-32
```

and this happens when player gets moved to -inf
```
2020-04-21 20:32:13: WARNING[Main]: input2.424246e-43
2020-04-21 20:32:13: WARNING[Main]: -0.013672 2.424246e-43
2020-04-21 20:32:13: WARNING[Main]: clipped X 3907.497803 0.000000 -inf
```

I'm pretty sure it didn't happen before 5.3 dev
